prometheus/cache.go

```diff
@@ -0,0 +1,236 @@
+	}
+
+	promCacheImpl struct {
+		cacheDuration          time.Duration
+		cacheAllRequestRates   map[string]map[string]timeInResult
+		cacheAppRequestRates   map[string]map[string]map[string]timeInOutResult
+		cacheNsSvcRequestRates map[string]map[string]timeInResult
+		cacheSvcRequestRates   map[string]map[string]map[string]timeInResult
+		cacheWkRequestRates    map[string]map[string]map[string]timeInOutResult
+		allRequestRatesLock    sync.RWMutex
+		appRequestRatesLock    sync.RWMutex
+		nsSvcRequestRatesLock  sync.RWMutex
+		svcRequestRatesLock    sync.RWMutex
+		wkRequestRatesLock     sync.RWMutex
+	}
+)
+
+func NewPromCache() PromCache {
+	kConfig := kialiConfig.Get()
+
+	cacheDuration := time.Duration(kConfig.ExternalServices.Prometheus.CacheDuration) * time.Second
+	promCacheImpl := promCacheImpl{
+		cacheDuration:          cacheDuration,
+		cacheAllRequestRates:   make(map[string]map[string]timeInResult),
+		cacheAppRequestRates:   make(map[string]map[string]map[string]timeInOutResult),
+		cacheNsSvcRequestRates: make(map[string]map[string]timeInResult),
+		cacheSvcRequestRates:   make(map[string]map[string]map[string]timeInResult),
+		cacheWkRequestRates:    make(map[string]map[string]map[string]timeInOutResult),
+	}
+	return &promCacheImpl
+}
+
+func (c *promCacheImpl) GetAllRequestRates(namespace string, ratesInterval string, queryTime time.Time) (bool, model.Vector) {
+	defer c.allRequestRatesLock.RUnlock()
+	c.allRequestRatesLock.RLock()
+
+	if nsRates, okNs := c.cacheAllRequestRates[namespace]; okNs {
+		if rtInterval, okRt := nsRates[ratesInterval]; okRt {
+			if queryTime.Sub(rtInterval.queryTime) < c.cacheDuration {
+				log.Tracef("[Prom Cache] GetAllRequestRates [namespace: %s] [ratesInterval: %s] [queryTime: %s]", namespace, ratesInterval, queryTime.String())
+				return true, rtInterval.inResult
+			}
+		}
+	}
+	return false, nil
+}
+
+func (c *promCacheImpl) SetAllRequestRates(namespace string, ratesInterval string, queryTime time.Time, inResult model.Vector) {
+	defer c.allRequestRatesLock.Unlock()
+	c.allRequestRatesLock.Lock()
+
+	if _, okNs := c.cacheAllRequestRates[namespace]; !okNs {
+		c.cacheAllRequestRates[namespace] = make(map[string]timeInResult)
+	}
+
+	c.cacheAllRequestRates[namespace][ratesInterval] = timeInResult{
+		queryTime: queryTime,
+		inResult:  inResult,
+	}
+	log.Tracef("[Prom Cache] SetAllRequestRates [namespace: %s] [ratesInterval: %s] [queryTime: %s]", namespace, ratesInterval, queryTime.String())
+}
+
+func (c *promCacheImpl) GetAppRequestRates(namespace string, app string, ratesInterval string, queryTime time.Time) (bool, model.Vector, model.Vector) {
+	defer c.appRequestRatesLock.RUnlock()
+	c.appRequestRatesLock.RLock()
+
+	if nsRates, okNs := c.cacheAppRequestRates[namespace]; okNs {
+		if appInterval, okApp := nsRates[app]; okApp {
+			if rtInterval, okRt := appInterval[ratesInterval]; okRt {
+				if queryTime.Sub(rtInterval.queryTime) < c.cacheDuration {
```

Comment:
Take care here if I'm correct, `queryTime` can be in past, so possibly lower than the query time stored previously in cache, hence diff being negative. In that case I guess we shouldn't hit cache?

Reply:
(I'm not talking for `GetAppRequestRates` specifically, but more in general when fetching metrics)

Reply:
I realize this cache is at the moment only for health-kind of queries, so it eliminates the case of fetching raw metrics in the past.... but still I think in the context of Graph Replay we do query health in the past, right? cc @jshaughn

Reply:
In which scenarios ? Perhaps I missed some use case.

Reply:
Actually, that's probably ok. `time.Sub` is not signed, so no matter that `queryTime` might be smaller than `rtInterval.queryTime`.
I'm just trying to figure out from code, not tested yet, in any case I guess it's important to test if health is correctly displayed during a Graph Replay as it may have some impacts there.

Reply:
> `if queryTime.Sub(rtInterval.queryTime) < c.cacheDuration`

Looking,  I don't think this is safe.  The cache only works if the queryTime is equal to or later than when the cache entry was stored.  queryTime.Sub will return a negative duration if queryTime is before rtInterval.queryTime.  And a negative duration would always be < c.cacheDuration.  I think you need something like, 
'''
`if !queryTime.Before(rtInterval.queryTime) && queryTime.Sub(rtInterval.queryTime) < c.cacheDuration`

Reply:
Oh, I think Github losses my response on this comment.
+1 on it.

Reply:
> queryTime.Sub will return a negative duration

Sure about that? Duration can be negative? That was my initial concern but it isn't clear in the documentation I've read

Reply:
@jotak Yes, I tested it yesterday just to make sure: For example:

```
func main() {
	start := time.Date(2000, 1, 1, 0, 0, 0, 0, time.UTC)
	end := time.Date(2000, 1, 1, 12, 0, 0, 0, time.UTC)

	difference := end.Sub(start)
	fmt.Printf("difference = %v\n", difference)

	difference = start.Sub(end)
	fmt.Printf("difference = %v\n", difference)
	d10s, _ := time.ParseDuration("10s")
	fmt.Printf("%s < 10s = %v", difference, (difference < d10s))
}

difference = 12h0m0s
difference = -12h0m0s
-12h0m0s < 10s = true
```